axlearn/common/flash_attention/common.py

```diff
@@ -86,3 +91,178 @@ def query_iterator_indices(block_mask_map: np.ndarray, *, padding: int = 0) -> K
+            dropout_rate: Dropout rate for attention probs.
+            interpret: Whether to use interpret mode for Pallas kernels.
+            tpu_block_size: Block size for TPU pallas kernels.
+            gpu_block_size: Block size for GPU pallas kernels.
+        """
+
+        is_decoding: bool = False
+        softmax_scale: float = 1.0
+        dropout_rate: float = 0.0
+        interpret: bool = False
+        tpu_block_size: int = 512
+        gpu_block_size: int = 128
+
+    def __init__(self, cfg):
+        super().__init__(cfg)
+        # Keep a typed copy of self.config.
+        self.cfg: BaseFlashAttention.Config = self.config
+
+    def name(self) -> str:
+        """Returns the class name."""
+        return self.__class__.__name__
+
+    def _log_unsupported(self, reason: str):
+        logging.warning("Not using %s because %s", self.name(), reason)
+        return False
+
+    def _check_block_size(self, *, query: Tensor, key: Tensor, block_size: int) -> bool:
+        q_seq_len = query.shape[1]
+        k_seq_len = key.shape[1]
+        if q_seq_len % block_size != 0 or k_seq_len % block_size != 0:
+            self._log_unsupported(f"{q_seq_len=} or {k_seq_len=} is not divisible by {block_size=}")
+            return False
+        return True
+
+    def is_supported(
+        self, query: Tensor, key: Tensor, value: Tensor, bias: BaseAttentionBias
+    ) -> bool:
+        """Returns whether the attention kernel supports the given configuration.
+
+        Args:
+            query: Query of shape [batch_size, target_length, num_heads, per_head_dim].
+            key: Key of shape [batch_size, source_length, num_kv_heads, per_head_dim].
+            value: Value of shape [batch_size, source_length, num_kv_heads, per_head_dim].
+            bias: Attention bias to apply.
+
+        Returns:
+            True if the current configuration is supported. False otherwise.
+
+        Raises:
+            ValueError: If the given configuration doesn't logically make sense, e.g. if the
+                shapes of q/k/v do not satisfy the requirement of a standard attention.
+        """
+        del bias
+        if key.shape != value.shape:
+            raise ValueError(f"Expects {query.shape=} to be equal to {key.shape=}")
+        if query.shape[0] != key.shape[0]:
+            raise ValueError(
+                f"Expects query batch size {query.shape[0]} to be equal to key batch size "
+                f"{key.shape[0]}"
+            )
+        if query.shape[-1] != key.shape[-1]:
+            raise ValueError(
+                f"Expects query head dim {query.shape[-1]} to be equal to key head dim "
+                f"{key.shape[-1]}"
+            )
+        if query.shape[2] % key.shape[2] != 0:
+            raise ValueError(
+                f"Expects query num heads {query.shape[2]} to be divisible by num key heads "
+                f"{key.shape[2]}"
+            )
+        return True
+
+    def __call__(
+        self,
```

Comment:
```suggestion
        self,
        *,
```
(If kwargs are prohibited somewhere, please leave a comment.)

Reply:
Left a comment